src/lib/utils/SynapseClient.ts

```diff
@@ -497,39 +458,74 @@ export const getFileHandleByIdURL = (
   )
 }
 
-export const getAsyncResultFromJobId = <T>(
-  urlRequest: string,
-  accessToken: string | undefined = undefined,
-  updateParentState?: any,
-): Promise<T> => {
-  return doGet(
-    urlRequest,
+/**
+ * Get a completed asynchronous job. Will refetch every 500ms until COMPLETE or FAILED.
+ * @param asyncJobId
+ * @param accessToken
+ * @param setCurrentAsyncResult - optional function that will receive the AsynchronousJobStatus object every time
+ *   it's fetched, including while it is in the "PROCESSING" state.
+ * @returns
+ */
+export const getAsyncResultFromJobId = async <TRequest, TResponse>(
+  asyncJobId: string,
+  responseBodyEndpoint: string,
+  accessToken?: string,
+  setCurrentAsyncResult?: (
+    result: AsynchronousJobStatus<TRequest, TResponse>,
+  ) => void,
+): Promise<AsynchronousJobStatus<TRequest, TResponse>> => {
+  let response = await doGet<AsynchronousJobStatus<TRequest, TResponse>>(
+    `/repo/v1/asynchronous/job/${asyncJobId}`,
     accessToken,
     undefined,
     BackendDestinationEnum.REPO_ENDPOINT,
   )
-    .then((resp: any) => {
-      // is this the job status?
-      if (resp.jobState && resp.jobState !== 'FAILED') {
-        updateParentState &&
-          updateParentState({
-            asyncJobStatus: resp,
-          })
-        // still processing, wait for a second and try again
-        return delay(500).then(() => {
-          return getAsyncResultFromJobId<T>(
-            urlRequest,
-            accessToken,
-            updateParentState,
-          )
-        })
-      }
-      // these must be the query results!
-      return resp
-    })
-    .catch(error => {
-      throw error
-    })
+  setCurrentAsyncResult?.(response)
+  while (response.jobState && response.jobState === 'PROCESSING') {
+    await delay(500)
+    response = await doGet<AsynchronousJobStatus<TRequest, TResponse>>(
+      `/repo/v1/asynchronous/job/${asyncJobId}`,
+      accessToken,
+      undefined,
+      BackendDestinationEnum.REPO_ENDPOINT,
+    )
+    setCurrentAsyncResult?.(response)
+  }
+
+  if (response.jobState === 'FAILED') {
+    /**
+     * While we technically already have the failure reason in the response, the HTTP response doesn't give a helpful error code (e.g. 403)
+     * that we can use for an error banner. We can get the HTTP code if we fetch the response body directly.
+     */
+    doGet<TResponse>(
+      responseBodyEndpoint,
+      accessToken,
+      undefined,
+      BackendDestinationEnum.REPO_ENDPOINT,
+    )
+  }
```

Comment:
I tried to make all requests use "/repo/v1/asynchronous/job/${jobId}" before realizing that we no longer get the HTTP status code for errors, which tells our ErrorBanner what to do. So if the async job fails, fetch it one last time and get the thrown error.